mission_control/src/mission.cpp

```diff
@@ -36,359 +36,44 @@
-
-    ROS_INFO("Stopping abort mission thread");
-    m_threadAbortMission->interrupt();
-    m_threadAbortMission->join();
-
-    // Clean up all of the behavior publishers and messages
-    ROS_INFO("Stopping and cleaning up active behaviors");
-    Behavior* cur_behavior = getNextAbortBehavior(true);
-    do
-    {
-      cur_behavior->stopBehavior();
-      cur_behavior = getNextAbortBehavior();
-    }
-    while (cur_behavior != NULL);
-  }
-}
-
-void Mission::addBehavior(Behavior* behavior) { m_behaviors.push_back(behavior); }
-
-Behavior* Mission::getNextBehavior(bool reset)
-{
-  static bool firsttime = true;
-  if (reset || (firsttime == true))
-  {
-    m_behavior_iterator = m_behaviors.begin();
-    firsttime = false;
-  }
-
-  if (m_behavior_iterator == m_behaviors.end())
-  {
-    return NULL;
-  }
-
-  return *m_behavior_iterator++;
-}
-
-void Mission::addAbortBehavior(Behavior* behavior) { m_aborts.push_back(behavior); }
-
-Behavior* Mission::getNextAbortBehavior(bool reset)
-{
-  static bool abortfirsttime = true;
-  if (reset || (abortfirsttime == true))
-  {
-    m_abort_behavior_iterator = m_aborts.begin();
-    abortfirsttime = false;
-  }
-
-  if (m_abort_behavior_iterator == m_aborts.end())
-  {
-    return NULL;
-  }
-
-  return *m_abort_behavior_iterator++;
-}
-
-void Mission::ProcessCorrectedPoseData(const pose_estimator::CorrectedData& data)
-{
-  if (m_current_behavior == NULL) return;
-
-  if ((GetState() != MissionState::ABORTING) && (GetState() != MissionState::EXECUTING)) return;
-
-  boost::mutex::scoped_lock callback_lock(m_mutCallbacks);
-  if (m_current_behavior->checkCorrectedData(data)) callBackTmr.stop();
-  callback_lock.unlock();
-}
-
-void Mission::SetState(MissionState state)
-{
-  boost::mutex::scoped_lock run_lock(m_MissionStateLock);
-  missionState = state;
-  run_lock.unlock();
-}
+using mission_control::Mission;
+using namespace mission_control;
+using namespace BT;
 
-Mission::MissionState Mission::GetState()
+Mission::Mission()
 {
-  MissionState temp;
-
-  boost::mutex::scoped_lock run_lock(m_MissionStateLock);
-  temp = missionState;
-  run_lock.unlock();
-
-  return temp;
+  missionStatus_ = BT::NodeStatus::IDLE;
+  missionFactory_.registerNodeType<WaypointBehavior>("Waypoint");
 }
 
-void Mission::AbortMissionWrapper(ros::NodeHandle nh)
-{
-  node_handle = nh;
-
-  AbortMission();
-}
+Mission::~Mission() {}
 
-void Mission::ExecuteMission(ros::NodeHandle nh)
+bool Mission::loadMission(const std::string &missionFullPath)
 {
-  if ((GetState() == MissionState::EXECUTING) || (GetState() == MissionState::ABORTING))
+  if (access(missionFullPath.c_str(), F_OK) != -1)
   {
-    ROS_WARN("Mission is already executing or aborting");
-    return;
+    missionFullPath_ = missionFullPath;
+    missionTree_ = missionFactory_.createTreeFromFile(missionFullPath_);
+    missionDescription_ = missionTree_.rootNode()->name();
+    printTreeRecursively(missionTree_.rootNode());
```

Comment:
@patapufate my earlier comment about *not* half-constructing `Mission`  instances still applies.

Reply:
@hidmic I'll need help with this